api/scripts/create-certification-center-memberships-from-organization-admins.js

```diff
@@ -52,18 +41,23 @@ function buildCertificationCenterMemberships({ certificationCenterId, membership
 }
 
 async function fetchCertificationCenterMembershipsByExternalId(externalId) {
-  const { id: certificationCenterId } = await getCertificationCenterByExternalId(externalId);
-  const memberships = await getAdminMembershipsByOrganizationExternalId(externalId);
-
-  return buildCertificationCenterMemberships({ certificationCenterId, memberships });
+  const certificationCenter = await getCertificationCenterByExternalId(externalId);
+  if (certificationCenter) {
+    const memberships = await getAdminMembershipsByOrganizationExternalId(externalId);
+    return buildCertificationCenterMemberships({ certificationCenterId: certificationCenter.id, memberships });
+  }
 }
 
 function prepareDataForInsert(rawExternalIds) {
   return Promise.all(_.map(rawExternalIds, ({ externalId }) => {
     return fetchCertificationCenterMembershipsByExternalId(externalId);
   }))
     .then((certificationCenterMemberships) => {
-      return _.union(...certificationCenterMemberships);
+      return _(certificationCenterMemberships)
+        .compact()
+        .flattenDeep()
+        .union()
+        .value();
```

Comment:
Petite explication  : 
Ici nous utilisons le fonctional programming permis par lodash, pour simplifier la lecture. 

Le tableau que nous avons à la base contient des tableaux de memberships à créer par centre de certification.

1. `compact` : nous permet d'enlever les valeurs `null`que peut nous retourner la méthode `fetchCertificationCenterMembershipsByExternalId`
2. `flattenDeep` : nous permet d'enlever l'imbrication de tableaux
3. `union` : nous permet de supprimer les doublons => il peut y avoir des doublons si en entrée nous avons plusieurs fois le même `externalId`

Nous obtenons donc un tableau de memberships à créer sans doublons

Reply:
Je ne comprends pas très bien, ce n'est pas ce que faisait (plus efficacement) le code qui est remplacé ? 🤔 

Je conviens que la combinaison `_.union(...certificationCenterMemberships)` est assez intense mais manifestement l'alternative ci-dessus nécessite quand même une explication 😉 

De plus je trouve le `flattenDeep` généralement suspect parce qu'il met justement à plat les tableaux à un niveau arbitraire de profondeur, ce qui est assez risqué si un élément qui ne devrait pas être aplati se trouve être identifié comme un tableau ; et ça n'est pas très parlant sur la profondeur de la liste attendue — alors qu'ici elle est fixe.

J'ai l'impression que s'il y a des problèmes dans l'existant, ils viennent plutôt de :
- utilisation de `.then` au lieu de `await` ;
- un problème de nommage : `certificationCenterMemberships` devrait s'appeler quelque chose comme `certificationCenterMembershipsLists` parce que c'est bien une liste de listes ;
- ambiguïté dans le retour de `fetchCertificationCenterMembershipsByExternalId` qui renvoie soit une liste de `memberships`, soit `undefined` — qui pourrait être rectifiée avec un `return []` dans le cas négatif (noter que ça marche actuellement parce que `union` ignore simplement ses arguments qui sont `undefined`, mais ça n'aide pas à la compréhension) ;
- un manque de traitement des données en entrées : s'il existe effectivement la possibilité de doublons dans les `externalId` en entrée, alors c'est sur cette liste en entrée qu'il faut appliquer un `_.uniq` !
- utilisation de `_.union` sur des objets qui ne sont pas des valeurs simplement comparables mais des structures complexes — je ne suis pas convaincu que la règle de comparaison par défaut appliquée par Lodash fonctionne dans ce cas !
- utilisation d'un `_.map` sur une collection de taille arbitraire, ce qui va déclencher l'utilisation d'un nombre délirant de connexions simultanées à la base de données — une meilleure technique est d'utiliser par exemple `bluebird.mapSeries` ou sinon `for`.

```suggestion
const externalIds = _.uniq(_.map(rawExternalIds, 'externalId'));
const certificationCenterMembershipsLists = await bluebird.mapSeries(externalIds, fetchCertificationCenterMembershipsByExternalId);
return certificationCenterMembershipsLists.flat();
```

Ou encore :

```suggestion
const externalIds = _.uniq(_.map(rawExternalIds, 'externalId'));
const result = [];
for(const externalId of externalIds) {
	result.push(...fetchCertificationCenterMembershipsByExternalId(externalId))
}
return result;
```